### PR TITLE
Implemented memory and CPU limits for LCOW.

### PR DESCRIPTION
**Pull request by notanaverageman**
Signed-off-by: Yusuf Tarık Günaydın <yusuf_tarik@hotmail.com>

**- What I did**
Added implementation of --memory and --cpu switches for LCOW.

**- How I did it**
Filled the Resources object in spec.Windows for both Windows and Linux by separating it into a new method. Also separated the code that fills the CPU and memory part of the hcsshim configuration and called it for both Windows and Linux.

**- How to verify it**
docker run -it --memory=2G --cpus=8 alpine sh -c "nproc && cat /proc/meminfo"

**- Description for the changelog**

Added memory and CPU limits for LCOW.

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by mlaventure** [comment]
You need to run `gofmt` and `goimport` on the modified files.

Also, could you add unit tests?

**Comment by yusuf-gunaydin** [comment]
I use VSCode and it automatically applies those tools on save. I cannot see any formatting or importing issues. If there is, can yo point me to it?

For unit testing, I can add test to https://github.com/moby/moby/blob/master/integration-cli/docker_cli_run_test.go, however, the docs says it is deprecated. I couldn't find a way to launch LCOW test from the new integration test suite.

By the way #37294 is very similar to this one and it is getting merged without any tests.

**Comment by thaJeztah** [comment]
Yeah, LCOW doesn't currently run in CI here (I think that's being worked on)

ping @johnstep @AntaresS ?



@yusuf-gunaydin looks like this needs a rebase (it shows there's a merge conflict); can you update the PR?

**Comment by lahma** [comment]
May I enquire the status of this? I tested @yusuf-gunaydin 's branch and it indeed resolved the memory starvation we had by allowing more memory to be assigned via command line.

**Comment by thaJeztah** [comment]
ping @johnstep @jhowardmsft PTAL

@yusuf-gunaydin looks like there's a merge-commit in this PR; could you rebase your branch, and squash the commits down to a single commit? Let me know if you need help doing so

**Comment by lahma** [comment]
FYI, the memory limit actually works, but CPU doesn't. the configuration is valid and goes to daemon as expected, but Windows doesn't yet handle the parameter properly. So when --cpu 20 is used, container always only sees 1.

**Comment by yusuf-gunaydin** [comment]
@thaJeztah I have applied the rebase. I assume the failing builds are unrelated as other pull requests fail with the same errors.

@lahma I tested with this command and result is correct:
```
docker run --rm -it --cpus=8 alpine sh -c "nproc"
8
```

This PR changes the limits on the created Hyper-V machine. I don't think there is an issue with Windows not supporting multiple CPUs with Hyper-V.

**Comment by jhowardmsft** [comment]
@beweedon just in case. As mentioned offline

**Comment by beweedon** [comment]
Running this on one of my machines seems to work fine, at least without your changes. Running `docker run --rm -it --cpus=4 alpine nproc` outputs 2, which is the default (again, this is without your changes). Also, I modified one of our tests that starts a Linux utility VM, and running `nproc` in the utility VM seems to respect the `ProcessorCount` value (i.e. setting `ProcessorCount` to 4 means `nproc` outputs 4). This might be a silly question, but there's more than one CPU available on the machine you're running on, right?

**Comment by lahma** [comment]
@beweedon this is a 32 core machine, but I found the fault by running your command. I think I was originally running --cpu-count that was logically named and had (Windows only) , it seems it's for Windows containers only and what I though this to be was the normal Windows parameter when hosting on Windows.

I could have sworn I did try the --cpus too, but I guess I'm wrong. Now I'm seeing the count I provided, --cpus 25 correctly reports the 25 cores in nprocs.

I'm sorry for making you chase a red herring.

**Comment by beweedon** [comment]
Oh, no problem at all! That's super confusing about those two parameter names. I'm glad it turned out not to be a bug :)

**Comment by richardgavel** [comment]
If all the discussion with this PR is related to CPU and memory is fine, can a PR be generated which just allows LCOW to be used with more than the default (which makes SQL Server for Linux impossible to use with LCOW)? This has been an issue for months.

File: daemon/oci_windows.go
Comment (jhowardmsft) [inline daemon/oci_windows.go]: Nit - should this just be called `SetResourcesInSpec` to be more obvious what it's doing?

File: libcontainerd/client_local_windows.go
Comment (jhowardmsft) [inline libcontainerd/client_local_windows.go]: Nit - Might be more obvious if this was called something more like `extractResourcesFromSpec`

**Review by lowenna — CHANGES_REQUESTED** [review]
I'm entirely in favour of these changes. In the current docker model where there's only one container per utility VM, this is the best that can really be done and mirrors what we do with Hyper-V WCOW containers. However, there was some pushback in previous PRs which did essentially exactly the same, IIRC from @thaJeztah. I can't recall what the pushback was. I believe there have been multiple attempts at this previously. 

However, this SGTM. @johnstep PTAL

**Comment by jhowardmsft** [comment]
Oh one more comment - these are containers on Windows, and the right thing to be controlling these things on Windows as it relates to the Utility VM for both WCOW and LCOW, is these two fields in the CLI: `---cpu-count` and `--cpu-percent`, as opposed to `--cpus`. This controls the Hyper-V scheduler.

From `docker run --help`

```
      --cpu-count int                  CPU count (Windows only)
      --cpu-percent int                CPU percent (Windows only)
```

**Comment by yusuf-gunaydin** [comment]
@jhowardmsft I have made the requested changes.

**Comment by jhowardmsft** [comment]
I don't think this is your change as it seems the same on master using Windows containers, but it's triggering against `--cpus`, as well as `--cpu-count`. I'm not sure when start came into being, but I wasn't the one who did most of the resource management implementation. Have verified it operates as expected for both memory and CPU, with the right document being passed to HCS. Example here for 5GB ram and 8 proc:

```
config={"SystemType":"container","Name":"aa690abd1c6b0894fc837d2a2e67212454ce305ff24d86f1008721cc7c956139","Owner":"docker","LayerFolderPath":"C:\\control\\lcow\\aa690abd1c6b0894fc837d2a2e67212454ce305ff24d86f1008721cc7c956139","Layers":[{"ID":"2e2ab5f1-6ea9-557b-af6c-c401a2a3db37","Path":"C:\\control\\lcow\\fb27f0fc7fbad278718cab180f3e0dcc238fc8509560b5db3e79804b0de1ec4a\\layer.vhd"}],"ProcessorCount":8,"MemoryMaximumInMB":5120,"HvPartition":true,"EndpointList":["F25FBCB5-9740-4B07-81E7-6B581DD1AACD"],"HvRuntime":{"ImagePath":"C:\\Program Files\\Linux Containers","LinuxInitrdFile":"initrd.img","LinuxKernelFile":"kernel"},"AllowUnqualifiedDNSQuery":true,"ContainerType":"linux","TerminateOnLastHandleClosed":true}
```

So LGTM. @johnstep & @thaJeztah PTAL.

```
PS E:\go\src\github.com\docker\docker> docker run -it --cpu-count=8 alpine sh -c "nproc"
8
PS E:\go\src\github.com\docker\docker> docker run -it --cpus=8 alpine sh -c "nproc"
8
PS E:\go\src\github.com\docker\docker> git branch
* lcow_limits
...snip....
PS E:\go\src\github.com\docker\docker> git log -1
commit 201746c83bca137ec338b71f7d0c650471dd2e3b (HEAD -> lcow_limits, yusuf/lcow_limits)
Author: Yusuf Tar<C4><B1>k G<C3><BC>nayd<C4><B1>n <yusuf_tarik@hotmail.com>
Date:   Sat Jun 16 10:01:35 2018 +0300

    Implemented memory and CPU limits for LCOW.

    Signed-off-by: Yusuf Tar<C4><B1>k G<C3><BC>nayd<C4><B1>n <yusuf_tarik@hotmail.com>
PS E:\go\src\github.com\docker\docker> docker version
...snip....
Server:
 Engine:
  Version:          0.0.0-dev
  API version:      1.39 (minimum version 1.24)
  Go version:       go1.11
  Git commit:       201746c83b
  Built:            09/28/2018 00:08:04
  OS/Arch:          windows/amd64
  Experimental:     true
PS E:\go\src\github.com\docker\docker>
```

**Review by lowenna — APPROVED** [review]

**Comment by jazzdelightsme** [comment]
It looks like this change is ready to go. @thaJeztah can this please be merged?

**Comment by Iristyle** [comment]
Also hoping that some form of this can be merged as we're running into problems running Linux containers on Windows due to OOM errors.

**Comment by MrLemur** [comment]
As far as I can see after compiling moby from source, I am able to get memory limits working even in Docker compose. Think we are just waiting on this to be merged.

**Comment by akiller** [comment]
We're also running into this trying to serve a Tensorflow model container, it needs ~5GB of memory.

I'm new to Docker so apologies if this is a stupid/irrelevant question here but if this pull request gets accepted how long would it be expected to take before it gets incorporated into Docker Enterprise? We're running Server 2019 which I believe is what it ships with.

**Comment by yusuf-gunaydin** [comment]
@jhowardmsft I have rebased again. Let's hope this would be the last one. 😇

**Comment by qacollective** [comment]
I've come here after a long voyage and would like to know how long after merge we should expect to see these changes in Docker for Windows.  It's amazing how hard it has been to simply stand up an SQL Server back end container plus a ASP.NET IIS front end container that can communicate!!  LCOW is currently being limited by the inability to have a container use more than 2Gb of RAM. Merge, go go go!

**Comment by jhowardmsft** [comment]
@johnstep @thaJeztah Please can you take a look? There are so many people waiting on this long-standing change.

**Comment by jhowardmsft** [comment]
ping @johnstep & @thaJeztah again.

**Comment by lecourtb** [comment]
When does this improvement will be available in a official Docker Desktop release ? I use the community version. Should I have to switch to the edge version ?

**Comment by jhowardmsft** [comment]
@orest-gulman Yes, that will be the case until this PR has been merged into moby/moby and docker-ce/ee takes the patch :/

**Comment by movainer** [comment]
@johnstep & @thaJeztah If i can add my 2c, can this go in?

**Comment by jhowardmsft** [comment]
@yusuf-gunaydin I can only apologise on behalf of other maintainers that this seems to keep being ignored. Unfortunately it needs another rebase. @thajeztah @johnstep @cpuguy83 what I can I do to convince other maintainers that this is required (for both local inprocess and the containerd work)? There’s no shortage of people who have indicated it’s a blocking issue for them, and it’s been ongoing for many many months now.

**Review by cpuguy83 — APPROVED** [review]
LGTM, but needs a rebase.

**Comment by jhowardmsft** [comment]
Thanks @cpuguy83. @yusuf-gunaydin If you can rebase, we should be able to get this merged. Thanks!

**Comment by yusuf-gunaydin** [comment]
@jhowardmsft Rebased again. Experimental build failed with a swarm test saying that context deadline exceeded. I think it's not related to this one, but I am not sure.

**Comment by jazzdelightsme** [comment]
Thank you @vdemeester! (and @yusuf-gunaydin!)

**Comment by orest-gulman** [comment]
> @orest-gulman Yes, that will be the case until this PR has been merged into moby/moby and docker-ce/ee takes the patch :/

Now we need to wait for an update docker ce?, thanks

**Comment by richardgavel** [comment]
Do you know if there's anything we can track on Docker's side to know they took the merge?

**Comment by Iristyle** [comment]
docker-ce repository has an `engine` directory at https://github.com/docker/docker-ce/tree/master/components/engine which contains the fork of the moby project that lives in https://github.com/docker/engine

docker-ce is pulling in commit https://github.com/docker/docker-ce/commit/c385bd144b1ee29edd45ec9a014336cd241c893b as of 2/6/2019, which, according to the commit, represents commit f8e29fdd683f8d1a79344df08342261c3991ad1b from this repo

f8e29fdd683f8d1a79344df08342261c3991ad1b includes the merge of PR https://github.com/moby/moby/pull/37296, so the nightly builds from master at https://master.dockerproject.com/windows/x86_64/docker.zip should now contain this feature (validating now)

For those interested in running / configuring nightlies, there are some instructions for installing this build (and configuring LCOW) that I've documented here:
https://github.com/puppetlabs/pupperware/blob/master/README-windows.md#install-the-docker-nightly-build

**Comment by albertony** [comment]
> I doubt this would be backported to 18.09 or earlier as LCOW remains an experimental feature.

It would be really nice if you actually would consider backporting, as LCOW in 19.03 seems to require Windows build 1809 (due to pull request #39108, with further explanation [here](https://github.com/moby/moby/issues/33850#issuecomment-478192332)), and some of us will be stuck with Windows build 1803 for some time (corporate managed computer). @thaJeztah have submitted a lot of backport pull requests on docker/engine recently, I would love to see this one included as well!

**Comment by waleed-alharthi** [comment]
> As far as I can see after compiling moby from source, I am able to get memory limits working even in Docker compose. Think we are just waiting on this to be merged.

Hi, a little late to the party, is this currently possible in compose? What tags to use?

**Pull request merged**
